runtime/src/main/java/org/corfudb/util/Utils.java

```diff
@@ -119,6 +117,80 @@ public static String toReadableId(UUID id) {
         return Long.toHexString((id.getLeastSignificantBits()) & 0xFFFF);
     }
 
+
+    /**
+     * Verify that the stripes are not empty and that each segment has one stripe.
+     * @param segments segments to validate
+     */
+    private static void validateSegments(List<Layout.LayoutSegment> segments) {
+        checkNotNull(segments);
+        checkArgument(!segments.isEmpty());
+        long previousSegmentEndAddress = 0;
+        for (Layout.LayoutSegment segment : segments) {
+            checkState(segment.getStart() == previousSegmentEndAddress);
+            previousSegmentEndAddress = segment.getEnd();
+            // only supported for chain replication
+            checkArgument(segment.getReplicationMode() == Layout.ReplicationMode.CHAIN_REPLICATION);
+            // Since stripping is not supported, we can assume only one stripe exists
+            checkArgument(segment.getStripes().size() == 1);
+            // A stripe cannot be empty
+            checkArgument(!segment.getStripes().get(0).getLogServers().isEmpty());
+        }
+        // The last segment (i.e open segment) end address is -1 (denoting infinity)
+        checkState(previousSegmentEndAddress == -1);
+    }
+
+    /**
+     * This method selects a head node: the first node (in the open segment) that also exists in all previous
+     * segments.
+     * @param layout layout to use to find the head node
+     * @return head node
+     * @throws IllegalStateException if no such node exists
+     */
+    private static String selectHeadNode(Layout layout) {
+        checkNotNull(layout);
+        validateSegments(layout.getSegments());
+        List<Layout.LayoutSegment> segments = layout.getSegments();
+
+        // select open segment (i.e. last segment)
+        Layout.LayoutSegment openSegment = segments.get(segments.size() - 1);
+        // only support chain replication
+        Layout.LayoutStripe openSegmentStripe = openSegment.getStripes().get(0);
+
+        if (segments.size() == 1) {
+            // Since there is only one segment we can return the first node in the stripe (i.e. head node)
+            return segments.get(0).getStripes().get(0).getLogServers().get(0);
+        }
+
+        // walk the chain of the last segment and find the first node that exists in all previous segments
+        List<Layout.LayoutSegment> reversedSegmentPrefix = segments.subList(0, segments.size() - 1);
+        checkState(!reversedSegmentPrefix.isEmpty());
+        // Reverse the prefix to simplify the indexing when walking the segments backwards (i.e. high to low)
+        Collections.reverse(reversedSegmentPrefix);
+
+        for (String node : openSegmentStripe.getLogServers()) {
+            boolean isNodeMissing = false;
+            for (Layout.LayoutSegment currentSegment : reversedSegmentPrefix) {
```

Comment:
If you are looking for a node that is present in all segments then no need to reverse the collection.
And a `double negative` conditions make the logic more error-prone.

For instance: !isNodeMissing - if the node is not missing, actually means - if the node present in all segments.
Also in the `for` cycle has to be simplified

Reply:
> If you are looking for a node that is present in all segments then no need to reverse the collection.

The algorithm isn't doing that. Although a head node will exist in all segments, merely existing in all segments doesn't make it a "head node". A head node is a node that's suppose to have the largest log prefix.

Reply:
> @Maithem Should we also fix the issue of the chain order, because that is the root cause of the this sequncer reconfiguration issue because it's based on the order assumption. And I think we have other places that assumes the order of the chain.

Given how state transfer works, I don't think its possible to maintain the same order of all nodes on all segments.

Consider the following scenario.

S1:{n2} S2:{n2, n3}  epoch = 5

then at epoch 6 we add n1

S1:{n2} S2:{n2, n3} S3:{n2, n3, n1}

At epoch 7 n1's state transfer completes
S1:{n2, n1} S2:{n2, n3, n1} S3:{n2, n3, n1}

At epoch 8, node two dies
S1:{n1} S2:{n3, n1} S3:{n3, n1}

As you can see here, node 1 is the tail for S3, but the head for S1. So its not possible to enforce order on the segments unless we schedule the state transfers properly, but I dont think that's a good idea. 

I'm changing my algorithm to a simpler approach where I select the head node of each segment and then aggregate the results.

Reply:
Actually, in the example above its clear that the sequencer state cannot be computed by selected only **one** node, so the current algorithm is incorrect. That's why we need to aggregate results from multiple nodes.

Reply:
Not sure why this replied to the wrong thread. @WenbinZhu see response above.

Reply:
@Maithem I see, this is an edge case in layout reconfiguration, n1 finishes state transfer earlier than n3.
I'll continue the review.